### PR TITLE
fixes #2113 - vmware: filter datastores and networks based on cluster

### PR DESCRIPTION
**Pull request by timogoebel**
This PR filters datastores and networks based on the selected cluster for vmware.
It needs https://github.com/fog/fog-vsphere/pull/35 

@angystardust : Would you mind testing this? Please pay special attention to compute profile handling as well. :-)

### CODE REVIEW

File: app/assets/javascripts/compute_resource.js [hunk 1]
```diff
@@ -244,27 +244,84 @@ function allocation_switcher(element, action) {
...
+  vsphereGetClusterItems(item, 'select.vmware_network', 'networkurl', __('Error loading networks: %s'), handleRequest);
+}
+
+function vsphereGetClusterItems(item, selector, datafield, errormsg, requesthandler) {
   var data = {cluster_id: $(item).val()};
-  var url = $(item).data('url');
+  var url = $(item).data(datafield);
+  var selectboxes = $(selector);
+  selectboxes.indicator_show();
   foreman.tools.showSpinner();
```
Comment (timogoebel) [inline app/assets/javascripts/compute_resource.js]: The spinner gets hidden too early, i.e. when the first request finishes. I'd prefer removing the global spinner completely and using only the "inline" spinners.

**Comment by angystardust** [comment]
Surely I will, @timogoebel !
Should I test against foreman-nighly or a patched 11.2 will also work?

**Comment by timogoebel** [comment]
@angystardust : Nightly is probably best, I haven't tried backporting this.

**Comment by xprazak2** [comment]
@timogoebel could you rebase pls?

**Comment by timogoebel** [comment]
@xprazak2: Thanks a lot for taking a look. Really appreciate it. I will rebase next week when i am back from my vacation. :-)

**Comment by xprazak2** [comment]
@timogoebel no problem, have a nice vacation!

**Comment by timogoebel** [comment]
@xprazak2 : Rebased. Thanks for taking a look.

**Comment by ohadlevy** [comment]
@timogoebel I assume the JS part needs to be converted to webpack directories? @tbrisker mind having a look on this?

**Comment by timogoebel** [comment]
@tbrisker : What does migrating this to webpack actually mean? I guess, it's more then just moving the files. I haven't got the hang of it.
I'd suggest, to first migrate this in a separate PR. I could rebase this PR afterwards.

Judging by the Redmine votes, this is a pr that the community wants. So I'd like to 🚢 this with 1.14.

**Comment by dLobatog** [comment]
@timogoebel Yeah basically moving the JS to that directory, and writing some tests (see webpack/**test** for examples). I haven't tried this PR yet. What did you mean by "to first migrate this in a separate PR"?

**Comment by timogoebel** [comment]
> What did you mean by "to first migrate this in a separate PR"?

@dLobatog : Write a pr to move the code to webpack / add the tests. And then rebase this pr on top of that. Do you think, this is a good idea?

**Comment by timogoebel** [comment]
@dLobatog : Besides all the 🐶 warnings, is this the right way to go?

**Comment by timogoebel** [comment]
Tried to fix all eslint problems.
How do can I get rid of the rest?

**Comment by xprazak2** [comment]
To remove 'no-undef' violations, something like `const __ = window.__;` should do the trick

**Comment by timogoebel** [comment]
I guess, i succeeded in migrating this to webpack. Any comments are truly appreciated.

**Comment by timogoebel** [comment]
I added some unit tests.
@dLobatog / @tbrisker : Could anyone of you take a look please?

**Comment by angystardust** [comment]
@timogoebel, there's no hope to see this patch in 1.13 GA, right?

**Comment by timogoebel** [comment]
@angystardust : Yes, unfortunately. Still waiting for a review.  I've been thinking if it would be wise to just extract the vsphere code to a plugin. But this also requires some work in Core, so it will not work with 1.13 as well. Sorry to disappoint you here. I know, you'd really like to have this feature.

**Comment by timogoebel** [comment]
Rebased.

**Comment by angystardust** [comment]
Hi @timogoebel still no chances to see this in 1.14GA?

**Comment by timogoebel** [comment]
@angystardust: Thanks for asking - again. I don't think, this will happen. Branching is in two or three weeks. I'm not sure exyctly. After react being introduced, we should change the code to use react to render the select fields. There are still some prequisites that need doing. We probably need Jed and a select2 react implementation. The code should be more straight forward, though.
Maybe @shlomizadok and I can team up and make a plan at OSMC to get both the scsi controller feature and this one done. I think, both features will be greatly appreciated by the community and it's really annoying not to get them done. Sorry, again.

**Comment by angystardust** [comment]
Is anybody still working on this patch? @timogoebel , since it's too late for the 1.15 GA, do you think there's hope to merge it in time for the 1.16? as usual I'm always available to test ;)

**Comment by timogoebel** [comment]
@angystardust : Good to know. I'll need something similar for cache invalidation. I'm pretty busy right now, but if I have some time this is definitely a priority.

**Comment by ohadlevy** [comment]
@timogoebel ping? :)

**Comment by timogoebel** [comment]
Maybe this would be something nice to do with react. @matanwerbner, what do you think? Basically it's just populating select fields with data from ajax.

**Comment by matanwerbner** [comment]
@timogoebel my initial thought is yes, we should try to avoid relying on jquery too heavily, and focus on quality FE work with react.
this is without being very familiar with the specific usecase, though.

**Comment by ohadlevy** [comment]
@timogoebel in other words - move the entire vm tab to react instead of just a small part of it? (controller/disks?)

**Comment by timogoebel** [comment]
> @timogoebel in other words - move the entire vm tab to react instead of just a small part of it? (controller/disks?)

@ohadlevy: Yeah. However I don't know if we have to go that far. This is a classic use case for react: A user changes one select field, another select field should get updated based on this value and ajax data. It's hard to do with jquery. I believe, react should be easier. And I say that even though I struggle with doing do a simple `npm i` for more than two hours.

@matanwerbner : What would be the best way to do an ajax call from a component? Is this where Redux comes into play?

**Comment by ohadlevy** [comment]
@timogoebel if you need a refresher on react and a bit on redux, i recently recorded this demo for another purpose but it might be useful.. https://bluejeans.com/s/vgQCW/

**Comment by matanwerbner** [comment]
@timogoebel yes, that is the idea.
in redux, components are in charge of firing actions, and updating themselfs to state changes.
the action fired triggers an ajax call done in a "action creator", the response of which changes the state via the reducers.

**Comment by angystardust** [comment]
Now that #4676 is merged, it would be great to pull this really useful feature in.
@timogoebel did you lose interest working on this one?

**Comment by timogoebel** [comment]
@angystardust : Not really, I just don't have the time. But if you want to give this a shot, go ahead. :-)

**Comment by ares** [comment]
is there any chance to resume work on this in foreseeable future? if not I might start learning react :-)

**Comment by timogoebel** [comment]
@ares: Sure, go ahead. If you have time to take this over, it'll be greatly appreciated. All the required vmware calls are already in this PR, they just have to be changes to support caching.
The hard part is just the UI of this. It might make sense to migrate the whole VMWare CR tab to react.

**Comment by ohadlevy** [comment]
> It might make sense to migrate the whole VMWare CR tab to react.

:+1: we already converted 60% of it, its probably not hard at all to do the remaining parts.

**Comment by timogoebel** [comment]
Good news everyone, I picked this PR up again. I don't know if I have time to finish it, but I got quite far in loading the datastores asynchronously.
Then we'd have to migrate the cluster select box to react and then we can easily connect both components.
This also needs some work in fog-vsphere, because querying datastores in a cluster is very slow.

**Comment by ares** [comment]
Fantastic news!

**Comment by timogoebel** [comment]
For reference: #6103 contains the first part we need for this. The data store and storage pod select field data is loaded asynchronously.

**Comment by chris1984** [comment]
Thanks again @timogoebel for fixing this one.

**Comment by timogoebel** [comment]
Closing in favor of #6442.

**Pull request closed without merging**